sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessorOptions.cs

```diff
@@ -52,6 +52,17 @@ public int PrefetchCount
         /// <value>true to complete the message automatically on successful execution of the message handler; otherwise, false.</value>
         public bool AutoCompleteMessages { get; set; } = true;
 
+        /// <summary>
+        /// Gets or sets whether the messages bodies use pooled byte arrays underneath in case of binary data.
+        /// When this option is enabled <see cref="ServiceBusReceivedMessage"/> cannot be safely kept for longer than the
+        /// lifetime of the scope of the <see cref="ServiceBusProcessor.ProcessMessageAsync"/> or
+        /// the <see cref="ServiceBusProcessor.ProcessSessionMessageAsync"/> event handler delegate.
+        /// The default value is false.
+        /// </summary>
+        ///
+        /// <value>true to pool message bodies; otherwise, false.</value>
+        public bool PoolMessageBodies { get; set; }
```

Comment:
What about `OptimizeBodyMemory` or similar? Trying to figure out a name that helps users understand why they may want to use this.

Reply:
I'm open for suggestions as long as we settle it in a discussion first and then change it once because changing it is slightly painful :D

I went for pooling because that is a well known concept in programming.

Reply:
As part of the public API, the name will be subject to architect review anyway...   if we feel that we're likely to move forward with the proposal, we should loop them in to avoid having to react to naming feedback after we brainstorm something.

Reply:
Is it best to wait for the highlevel blessing before I continue to invest in polishing this PR?

Reply:
Yes, I think that makes sense. I will try to bring it up this week.

Reply:
How about something wordy like:

- ScopeBinaryMessageBodyToProcessMessageDelegate
- ScopeBinaryMessageBodyToProcessMessageDelegateForHighthroughputScenarios

Reply:
Any updates?

Reply:
Yeah, sorry for the delay. The feedback we got when discussing this internally was that it may be too early to make this kind of optimization that changes the API and programming model, especially since there are performance bottlenecks in the underlying AMQP library that we would want to try to address first. For instance, there is all the string and ByteBuffer allocations that are even larger than the allocations involving the message body copying. Also, thanks to your continued efforts, we are making progress improving performance without changing the API. So at least for the time being, we'd like to continue making these performance improvements to the SDK, and start investigating potential improvements to the AMQP library.

Reply:
I'll add some thoughts later today. Please don't close it just yet

Reply:
Of course - wasn't planning on closing this 😄

Reply:
> The feedback we got when discussing this internally was that it may be too early to make this kind of optimization that changes the API and programming model, especially since there are performance bottlenecks in the underlying AMQP library that we would want to try to address first. For instance, there is all the string and ByteBuffer allocations that are even larger than the allocations involving the message body copying.

While I agree the underlying AMQP library should be optimized as well, and I'm also willing to help a bit there I have one concern with this argument. The concern I have is the design approach of this SDK. For convenience reasons the design choice that has been made is to copy whatever comes from AMQP into the annotated message and the body (I'm not disagreeing or shooting down this, I agree it is good for the majority of the users). 

So by design regardless of whether the AMQP library is optimized or not this library will have to copy the memory to release the underlying already buffered memory. So based on that having the possibility to opt out from always creating new memory vs having a predictable and stable memory load for high-performance scenarios is something super valuable to have regardless of how the AMQP library is optimized since I don't think you would ever bring back explicit releasing/disposing of the ServiceBusReceivedMessages back to the majority of the users. Taking this into account I think comparing AMQP memory traffic with the memory traffic of the design choice from copying AMQP buffers into the higher level primitives of this library is comparing to orthogonal concerns which shouldn't be compared.

Reply:
Well the copying the memory part is a result of the design of the AMQP library, since the body uses rented memory. If this weren't the case, then I don't think we would copy the memory.

Reply:
But that's the thing here. Any decent modern memory optimized library will want to use for larger memory chunks and explicit rent and release cycle. So if you expose that memory directly to the customer they become responsible for releasing. Otherwise if that is not a safe tradeoff you are willing to make you need to copy. I don't think you can avoid that and from that perspective Amqp lib does the right thing

Reply:
> So based on that having the possibility to opt out from always creating new memory vs having a predictable and stable memory load for high-performance scenarios is something super valuable to have regardless of how the AMQP library is optimized since I don't think you would ever bring back explicit releasing/disposing of the ServiceBusReceivedMessages back to the majority of the users.

But if the underlying AMQP library continues to allocate memory (which I think it does), then even if the SDK doesn't we would still not have stable memory load.

Reply:
Here is what I was seeing in a snapshot from processing 10K messages:
![dotmemory1](https://user-images.githubusercontent.com/54595583/115924244-52d6de00-a434-11eb-908a-752792691a8b.png)

String, ByteBuffer, and AmqpSymbol are all from the AMQP library.

Reply:
Correct the AMQP library continues to allocate and also unnecessarily box and unbox which is a problem. Yet again I still think these allocations are orthogonal to the allocations in the ServiceBus SDK

FYI some allocations and boxing could be removed if @eerhardt  finishes the PR started https://github.com/Azure/azure-amqp/pull/170 which would be super nice.

I've also tried to get rid of some of the custom buffer pooling underneath in this PR

https://github.com/Azure/azure-amqp/pull/182

Reply:
I probably won't have time in the near future to finish https://github.com/Azure/azure-amqp/pull/170. If someone wants to take it over and complete it for the other primitive types, that would be alright with me.

Reply:
I might be able to contribute a bit but can't promise much. It is a great PR by the way! I went down the road of using generics recently until I discovered yours which achieves similar benefits with less complexity